### PR TITLE
SWATCH-401: Apply the new candlepin facts to the cores formula

### PR DESCRIPTION
**Pull request by Sgitario**
Jira issue: [SWATCH-401](https://issues.redhat.com/browse/SWATCH-401)

## Description
- If system profile "threads per core" is set, use it instead of the default 2.0
- Else if system profile "number of cpus" is set, use the formula provided in the epic description
- Else use 2.0 as the default value as it was being done

This has been configured to be applied to all products. However, if we detect that the new threads per core value is misleading, we can disable it using the env property `USE_CPU_SYSTEM_FACTS_TO_ALL_PRODUCTS=false`.

Moreover, a warning message will be printed when the threads per core is other than 2.0 for troubleshooting purposes. 

## Testing
Added tests to cover this change.
QE needs to add new test coverage to verify the e2e feature.

### CODE REVIEW

**Comment by Aurobinda55** [comment]
> The test failures are caused because the number of cpus is set and the resulting threads per core is other than 2. I'm not sure if this is because of wrong test data, or the assertion needs to be updated or whether we should get rid of the formula that uses the number of cpus. @kahowell @Aurobinda55

@Sgitario I am getting different results and not match the calculation mentioned in the card - 

case -1 -
Number of sockets = 5
Cores_per_socket = 4
Cpu=8
threads_per_core=3
if I go with the formula then 
Cores= sockets(5)*core_per_sockets(4)/3 which is 5. But I am getting 7 
Case-2 -

Number of sockets = 5
Cores_per_socket = 4
Cpu=24
threads_per_core=2
new threads_per_core=number_of_cpus / (number_of_sockets * cores_per_socket)  which is 24/(5*4) = 1
cores= 5*4/1 which is 20, but I am getting 10 

Could you please confirm if I am missing anything ? 

About the tests failing, yes, something missing due to this latest change. I will take a look once the manual verification is completed.

**Comment by Sgitario** [comment]
There are a few failing tests that are related to these changes. Yet I'm unsure if the tests were

> threads_per_core

As long as the threads_per_core is provided, this is the number that rules. 

In case 1, you use `threads_per_core=3`, so the vCPUs (sockets(5)*core_per_sockets(4))/threads_per_core(3) = 6.667 which is 7, not 5. 
In case 2, the same, you use `threads_per_core=2`, so the vCPUs  (sockets(5)*core_per_sockets(4))/threads_per_core(2) which is 10, not 5.

**Review by kahowell — CHANGES_REQUESTED** [review]

**Comment by Aurobinda55** [comment]
> There are a few failing tests that are related to these changes. Yet I'm unsure if the tests were
> 
> > threads_per_core
> 
> As long as the threads_per_core is provided, this is the number that rules.
> 
> In case 1, you use `threads_per_core=3`, so the vCPUs (sockets(5)*core_per_sockets(4))/threads_per_core(3) = 6.667 which is 7, not 5. In case 2, the same, you use `threads_per_core=2`, so the vCPUs (sockets(5)*core_per_sockets(4))/threads_per_core(2) which is 10, not 5.

@Sgitario thanks for the explanation. I have still one doubt about the second case, threads_per_core=2 which is a default value, and the number of cupus=24, will it not use the formula to calculate the threads_per_core like mentioned in https://github.com/RedHatInsights/rhsm-subscriptions/pull/2859/files#diff-cc5b91fb352cbf09c7be01446edf9df8c25ffa879372fca77ca6368efd9fe21cR619 ?

**Comment by Sgitario** [comment]
@kahowell PR updated. I've added the property in the template as well.

**Comment by Aurobinda55** [comment]
> > > There are a few failing tests that are related to these changes. Yet I'm unsure if the tests were
> > > > threads_per_core
> > > 
> > > 
> > > As long as the threads_per_core is provided, this is the number that rules.
> > > In case 1, you use `threads_per_core=3`, so the vCPUs (sockets(5)*core_per_sockets(4))/threads_per_core(3) = 6.667 which is 7, not 5. In case 2, the same, you use `threads_per_core=2`, so the vCPUs (sockets(5)*core_per_sockets(4))/threads_per_core(2) which is 10, not 5.
> > 
> > 
> > @Sgitario thanks for the explanation. I have still one doubt about the second case, threads_per_core=2 which is a default value, and the number of cupus=24, will it not use the formula to calculate the threads_per_core like mentioned in https://github.com/RedHatInsights/rhsm-subscriptions/pull/2859/files#diff-cc5b91fb352cbf09c7be01446edf9df8c25ffa879372fca77ca6368efd9fe21cR619 ? if not is there any case when the formula to calculate thread_per_core should trigger?
> 
> Did you set `threads_per_core=2` in the test case? If so, it will be used and not trigger the formula. The formula will be used when threads_per_core is unset (or it has 0 value).

Thanks, @Sgitario for the confirmation...  I will take a look today and update the PR.

File: swatch-system-conduit/src/main/java/org/candlepin/subscriptions/conduit/InventoryController.java
Comment (Sgitario) [reply swatch-system-conduit/src/main/java/org/candlepin/subscriptions/conduit/InventoryController.java:179]: This is only meant for troubleshooting purposes for testing these changes. I will delete this change once everything looks good for QE. 
Moving to draft PR to not merge this.

Comment (Sgitario) [reply swatch-system-conduit/src/main/java/org/candlepin/subscriptions/conduit/InventoryController.java]: PR updated by using debug level.

**Comment by Aurobinda55** [comment]
@Sgitario I have checked the below test scenarios and everything looks good. There is only one problem that we discussed on the chat related to the number_of_cpus auto-assignment. I tried to find the root cause but couldn't so raised [SWATCH-2028](https://issues.redhat.com/browse/SWATCH-2028) to troubleshoot more and track the issue. Also, created [PR-529](https://gitlab.cee.redhat.com/insights-qe/iqe-rhsm-subscriptions-plugin/-/merge_requests/529
) to fix the test failures due to the new core formula. Approve the PR based on the below test scenarios. 
```
Case-1- sockets=5
        cores_per_socket=4
        number_of_cpu=4
Output:
Cores= 100
******************************
Case-2- Initial cores=100
        sockets=5
        cores_per_socket=4
        number_of_cpu=4
        threads_per_core=3    
Output:
Cores= 107
*****************************
Case-3:
Initial cores=107
        sockets=4
        cores_per_socket=4
        number_of_cpu=4
        threads_per_core=0
output:
Cores= 171
*****************************
Case-4:
Initial cores=171
        sockets=4
        cores_per_socket=4
        number_of_cpu=0
        threads_per_core=0
Output-
Cores=179

***************************

Case-5:
Initial cores=179
        sockets=4
        cores_per_socket=4
        number_of_cpu=0
Output:
Cores= 187

***************************

Case -6:
Initial cores=187
        sockets=0
        cores_per_socket=4
        number_of_cpu=0
Output:
Cores= 187

*************************

Case -7:
Initial cores=187
        sockets=4
        cores_per_socket=0
        number_of_cpu=0
Output:
Cores= 187

```

**Review by Aurobinda55 — APPROVED** [review]

**Review by kahowell — APPROVED** [review]

**Comment by Sgitario** [comment]
We should wait until the mentioned MR is merged before merging this one. Otherwise, all the following PRs will start failing. 
Since this feature has higher priority, @Aurobinda55 can you make sure that the MR is merged as soon as the changes are ok?

**Pull request merged**
